### ISSUE TITLE
Unexpected checkPermission() results 

### ISSUE THREAD

**Issue by averstuyf** [Baseflow/flutter-geolocator#639]
## 🐛 Bug Report

Android 11, API 30 has four Location Permissions a user can choose from:

- Allow all the time
- Allow only while using the app
- Ask every time
- Deny

and the special one that only exists if the app never had any Location Permission set:

- None

These seem to map to LocationPermission enum as follows:

- Allow all the time -> always
- Allow only while using the app -> whileInUse
- Ask every time -> denied
- Deny -> denied
- None -> denied

> Note that 'deniedForever' is never used and three options are represented by the same LocationPermission.denied.

> Also 'Ask every time' sometimes returns 'deniedForever', if set through the system permission window by choosing 'Only this time'.

The following logic/behaviour is expected before the app uses location data:

- Allow all the time -> use location data
- Allow only while using the app -> use location data
- Ask every time -> ask permission through Geolocator.requestPermission()
- Deny -> ask permission through Geolocator.openAppSettings()
- None -> ask permission through Geolocator.requestPermission()

The problem exists where Ask every time, Deny and None all map to denied. So no differentiation in case can be made.

### Expected behavior

A more fine grained mapping of Android 11, API 30, location permissions. Ideally one-to-one.

### Reproduction steps

Look at the result of a call to Geolocator.checkPermission() running on Android 11, API 30. The example app worked for this, without any modifications.

### Configuration

Google Pixel 2 XL, updated. Android 11, API 30. Flutter 1.22.5. Project target API 30.

**Version:** 1.1.13

**Platform:** 
- [ ] :iphone: iOS
- [x] :robot: Android

**Comment by averstuyf** [comment]
Bonus side-effect: 

Per the documentation, asking for coordinates should bring up the permission screen if none were chosen before. 

That might no longer work due to the problem, described above, where choosing 'Ask every time' in the App Location Permissions screen provide unreliable results.

**Comment by mvanbeusekom** [comment]
Hi @averstuyf, 

Thank you for reporting this issue, I will look into it straight away. But first I would like to list what is expected (or at least how I intended it to work) on Android 11,  API 30 (there is a slightly different behaviour on earlier API levels):

- Selecting "Allow all the time" should result in a `LocationPermission.always`;
- Selecting "Allow only while in using the app" should result in `LocationPermission.whileInUse`;
- Selecting "Ask every time" should result in `LocationPermission.whileInUse` during the same session, when the App is killed and started again calling the `checkPermission` method should result in a `LocationPermission.denied`;
- Selecting "Deny" for the first time the `checkPermission` should result in a `LocationPermission.denied`;
- Selecting "Deny" after a second time requesting permissions (while first time was denied) the result should be `LocationPermission.deniedForever` (starting from API 30, Android no longer offers a separate checkbox where one can select "Don't ask me again", instead Android assumes that a second deny is considered as "denied forever");
- Lastly the "None" situation you describe doesn't exists in Android. Android assumes that permissions are denied unless specifically indicated otherwise by the device's user. So if you check permissions before you requested them, Android will return that permissions are denied.

I will check if this is indeed how things are working and adjust the code if necessary. Will keep you updated here.

**Comment by mvanbeusekom** [comment]
@averstuyf I have run some tests and the permissions behave as mentioned above with the exception of setting the permission to "Ask every time" when the last state was `deniedForever`. To reproduce this situation I took the following steps with the Geolocator example App:

1. Make sure you start fresh (meaning make sure you remove the example App from your test device);
2. Start the App and hit the `getCurrentPosition` button;
3. When ask for permissions select the "Deny" option;
4. Hit the `getCurrentPosition` button once more;
5. Again select the "Deny" option;
6. You should now see that the permission status returned == `deniedForever`;
7. Next leave the App and go into the location permission settings (long press the app icon -> App info -> Permissions -> Location);
8. Now select "Ask every time" and leave the settings;
9. Start the "geolocator_example" app and press "getPermissionStatus";
10. This should return the `denied` permission but instead returns `deniedForever`;

This situation is caused due to the fact that we store a boolean in the shared preferences when the user denied permissions and the Android `shouldShowRequestPermissionRationale` method returns `false`. Before API 30 this was a reliable way to determine the user hit the `don't ask again` checkbox however since API 30 and the introduction of the `Ask every time` permission this is causing issues. Remember Android always starts out  denying permissions until the user explicitly gave permissions. Meaning the `Ask every time` permission simply resets the permission to `denied`, now unfortunately the `shouldShowRequestPermissionRationale` also keeps returning false and we have the boolean set from last time we ran the App, so the Geolocator thinks permissions are still `deniedForever`.

I will need some time to consinder this to come up with the best solution, unfortunately Android doesn't have an API or method which allows us to verify if permissions are denied forever of if they have been reset by the users. Android will only return `granted` or `denied` when checking permissions. If we request permissions Android will internally determine if it should display the permission dialog (and return `granted` or `denied` depending on the users selection) or not and if will simply return `denied`.

So for now I will leave this issue open and hopefully come up with some solution soon. Of course feedback is more then welcome!

**Comment by averstuyf** [comment]
Thanks for looking into this. 

This most Location Permission revision seems to have increased complexity to the point where it does not make sense anymore. Sending users to browse around in settings, considering many have trouble with basic smartphone functionality already, is feels productive.

Together with other changes Google has pushed through recently, time is wasted. I'm going to advice a wait-and-see approach here, while diverting resources to Apple development.

**Comment by willbryant** [comment]
I found the discussion on https://stackoverflow.com/questions/32347532/android-m-permissions-confused-on-the-usage-of-shouldshowrequestpermissionrati helpful.

In particular, the accepted answer sounds like a better approach now than the advice Google gave to call the 'should' method first (which then motivated the use of a stored pref as you have done).

**Comment by mvanbeusekom** [comment]
This issue should be solved in version 7.0.0-nullsafety.7. It required a rework of the permission model on Android which explains the update of the major version number.

I tried to document the changes as good as possible in this wiki page: https://github.com/Baseflow/flutter-geolocator/wiki/Breaking-changes-in-7.0.0#android-permission-update

I will close this issue for now but please let me know if things are not clear or if I overlooked something. I would be happy to reopen the issue and have another look.

**Comment by vpuonti** [comment]
Hello!

I am an Android dev and stumbled upon this issue when researching Android's location permissions, namely getting info on whether or not the permission has been denied forever or not. 
For the longest time I used the same hacks as everyone else (`if(!isPermissionGranted && !shouldShowRationale)` in `registerForActivityResult()` callback, but with the upcoming Android 12 location permission rework I noticed something was wrong. 

I noticed that on Android 11 and Android 12 the user can dismiss the native permission dialog by clicking outside the dialog or pressing back button. When the dialog is dismissed `registerForActivityResult()` callback gets called and the *values of `isPermissionGranted` and `shouldShowRationale` are`false`*.  

I also checked the result of cancellation in `onRequestPermissionsResult()` to verify that this was not a bug in the ActivityResult API with the following code:

Code:
```kotlin
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        Timber.d("permissions: ${permissions.toList()} \ngrantResults: ${grantResults.toList()}")
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
```
Output:
```
D/AppActivity: permissions: [android.permission.ACCESS_COARSE_LOCATION, android.permission.ACCESS_FINE_LOCATION] 
    grantResults: [-1, -1]
```

To quote the [documentation](https://developer.android.com/reference/android/app/Activity#onRequestPermissionsResult(int,%20java.lang.String[],%20int[])):

> Note: It is possible that the permissions request interaction with the user is interrupted. In this case you will receive empty permissions and results arrays which should be treated as a cancellation. 

It would seem like one of these are true: 
* there is a flaw in how Android 11 & Android 12 deal with permission dialog cancellation (manually cancelling the permission dialog does not count towards `shouldShowRequestPermissionRationale()`)
* There is a flaw in `onRequestPermissionsResult()` documentation 
* I am misunderstanding what the concept of "interruption" means in the context of permission request flow

If it doesn't already exist, I will open an issue on Android Issue Tracker about this to (hopefully) get some clarification. 
Until then, it would seem like there are a few options for developers: 
* handle `permissionDeniedForever` also when the user dismisses the permission dialog on the first time it is prompted
* completely dismiss `permissionDeniedForever` as a concept (NOT GOOD)
* Conform to Google's *inflexible* vision of how permissions should be asked https://developer.android.com/training/permissions/requesting#workflow_for_requesting_permissions

If there is something I have missed please let me know!
